userspace/libsinsp/metrics_collector.cpp

```diff
@@ -0,0 +1,761 @@
+/*
+Copyright (C) 2024 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#ifdef __linux__
+
+#include <libsinsp/sinsp_int.h>
+#include <libsinsp/metrics_collector.h>
+#include <cmath>
+#include <sys/times.h>
+#include <sys/stat.h>
+#include <re2/re2.h>
+
+static re2::RE2 s_libs_metrics_units_suffix_pre_prometheus_text_conversion("(_kb|_bytes|_mb|_perc|_percentage|_ratio|_ns|_ts|_sec|_total)", re2::RE2::POSIX);
+static re2::RE2 s_libs_metrics_units_memory_suffix("(_kb|_bytes)", re2::RE2::POSIX);
+static re2::RE2 s_libs_metrics_units_perc_suffix("(_perc)", re2::RE2::POSIX);
+static re2::RE2 s_libs_metrics_banned_prometheus_naming_characters("(\\.)", re2::RE2::POSIX);
+
+static const char *const sinsp_stats_v2_resource_utilization_names[] = {
+	[SINSP_RESOURCE_UTILIZATION_CPU_PERC] = "cpu_usage_perc",
+	[SINSP_RESOURCE_UTILIZATION_MEMORY_RSS] = "memory_rss_kb",
+	[SINSP_RESOURCE_UTILIZATION_MEMORY_VSZ] = "memory_vsz_kb",
+	[SINSP_RESOURCE_UTILIZATION_MEMORY_PSS] = "memory_pss_kb",
+	[SINSP_RESOURCE_UTILIZATION_CONTAINER_MEMORY] = "container_memory_used_bytes",
+	[SINSP_RESOURCE_UTILIZATION_HOST_CPU_PERC] = "host_cpu_usage_perc",
+	[SINSP_RESOURCE_UTILIZATION_HOST_MEMORY] = "host_memory_used_kb",
+	[SINSP_RESOURCE_UTILIZATION_HOST_PROCS] = "host_procs_running",
+	[SINSP_RESOURCE_UTILIZATION_HOST_FDS] = "host_open_fds",
+	[SINSP_STATS_V2_N_THREADS] = "n_threads",
+	[SINSP_STATS_V2_N_FDS] = "n_fds",
+	[SINSP_STATS_V2_NONCACHED_FD_LOOKUPS] = "n_noncached_fd_lookups",
+	[SINSP_STATS_V2_CACHED_FD_LOOKUPS] = "n_cached_fd_lookups",
+	[SINSP_STATS_V2_FAILED_FD_LOOKUPS] = "n_failed_fd_lookups",
+	[SINSP_STATS_V2_ADDED_FDS] = "n_added_fds",
+	[SINSP_STATS_V2_REMOVED_FDS] = "n_removed_fds",
+	[SINSP_STATS_V2_STORED_EVTS] = "n_stored_evts",
+	[SINSP_STATS_V2_STORE_EVTS_DROPS] = "n_store_evts_drops",
+	[SINSP_STATS_V2_RETRIEVED_EVTS] = "n_retrieved_evts",
+	[SINSP_STATS_V2_RETRIEVE_EVTS_DROPS] = "n_retrieve_evts_drops",
+	[SINSP_STATS_V2_NONCACHED_THREAD_LOOKUPS] = "n_noncached_thread_lookups",
+	[SINSP_STATS_V2_CACHED_THREAD_LOOKUPS] = "n_cached_thread_lookups",
+	[SINSP_STATS_V2_FAILED_THREAD_LOOKUPS] = "n_failed_thread_lookups",
+	[SINSP_STATS_V2_ADDED_THREADS] = "n_added_threads",
+	[SINSP_STATS_V2_REMOVED_THREADS] = "n_removed_threads",
+	[SINSP_STATS_V2_N_DROPS_FULL_THREADTABLE] = "n_drops_full_threadtable",
+	[SINSP_STATS_V2_N_MISSING_CONTAINER_IMAGES] = "n_missing_container_images",
+	[SINSP_STATS_V2_N_CONTAINERS] = "n_containers",
+};
```

Comment:
Can we `static_assert` that the size of `sinsp_stats_v2_resource_utilization_names` matches `SINSP_MAX_STATS_V2` to avoid future issues?
Same with below names arrays.